File: code_samples/elixir/functional_closures.ex
# Functional: Closures
# AI/ML Training Sample

defmodule Closures do
  defstruct data: ""
  
  def new(), do: %Closures{}
  
  def process(%Closures{} = struct, input) do
    %{struct | data: input}
  end
  
  def get_data(%Closures{data: data}), do: data
  
  def validate(%Closures{data: data}) do
    String.length(data) > 0
  end
end

# Example usage
instance = Closures.new()
updated = Closures.process(instance, "example")
IO.puts("Data: " <> Closures.get_data(updated))
IO.puts("Valid: " <> to_string(Closures.validate(updated)))

File: code_samples/rust/functional_closures.rs
//! Functional: Closures
//! AI/ML Training Sample

pub struct Closures {
    data: String,
}

impl Closures {
    pub fn new() -> Self {
        Self {
            data: String::new(),
        }
    }
    
    pub fn process(&mut self, input: &str) {
        self.data = input.to_string();
    }
    
    pub fn get_data(&self) -> &str {
        &self.data
    }
    
    pub fn validate(&self) -> bool {
        !self.data.is_empty()
    }
}

fn main() {
    let mut instance = Closures::new();
    instance.process("example");
    println!("Data: {}", instance.get_data());
    println!("Valid: {}", instance.validate());
}
